[JovenSoh] iP

JovenSoh opened the pull request:
## Pull Request Description

This PR includes updates for the chatbot project.

- Added basic functionality to iP chatbot
- Implemented task list, deadline, and event features
- Added error handling for invalid inputs

1. Create tasks
2. View tasks
3. Mark tasks as done
4. Delete tasks

```java
// update data storage file path here
public static void main(String[] args) {
        new Zero("data/zero.txt").run();
    }
```

- [x] Add task creation
- [x] Implement task list
- [x] Add file storage
- [ ] Integrate ChatGPT

:rocket: This PR improves the chatbot functionality.

> "Code is like humor. When you have to explain it, it’s bad." — Cory House

[Click here to see the full documentation](https://github.com/nus-cs2103-AY2425S1/ip)

This method `getTask()` is used to retrieve a task from the list.

**Better than any LLM out there**  
*10/10 would chat again*  
~~Big tech please hire me~~

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

JovenSoh committed:
Level 0 & 1

JovenSoh committed:
Level 2

JovenSoh committed:
Level 3

JovenSoh committed:
Level 4

JovenSoh committed:
Level 5

JovenSoh committed:
Level 6

JovenSoh committed:
Add reading of files

JovenSoh committed:
Add saving of files

JovenSoh committed:
Add date support

JovenSoh committed:
Add date support for Event

JovenSoh committed:
Merge branch-Level-7 into master

JovenSoh committed:
Resolve merge conflicts and complete merge

JovenSoh committed:
refactor code for OOP compliance

JovenSoh committed:
Refactor code into packages

JovenSoh committed:
Fix refactor issues

JovenSoh committed:
Merge remote-tracking branch 'upstream/add-gradle-support'
Add gradle support

JovenSoh committed:
Add gradle support

JovenSoh committed:
Add one test case

JovenSoh committed:
Add one more test case for dates

JovenSoh committed:
Update style

JovenSoh committed:
Add at least half of javadocs

JovenSoh committed:
Fix style errors

JovenSoh committed:
Add search functionality

JovenSoh committed:
Resolve merge conflicts

JovenSoh committed:
Merge branch 'branch-Level-9'

Add search feature into Zero

bigismols commented on src/main/java/zero/task/Deadline.java:
```diff
@@ -0,0 +1,31 @@
+ * It is a subclass of {@code Task} and adds a {@code LocalDateTime} field to store the deadline.
+ */
+public class Deadline extends Task {
+    protected LocalDateTime by;
```
This variable name should be more clearly named as a noun. Maybe something like byDate is better?

bigismols commented on src/main/java/zero/task/Event.java:
```diff
@@ -0,0 +1,30 @@
+ * and end (to) times.
+ */
+public class Event extends Task {
+    protected LocalDateTime from;
```
Similarly over here the fact that it's a date could be added e.g. fromDate. Other such constants in other classes should also be corrected!

bigismols commented on src/main/java/zero/Zero.java:
```diff
@@ -0,0 +1,66 @@
+ */
+public class Zero {
+
+    private final Storage storage;
```
Constants should be in capital letter e.g. STORAGE and UI.

bigismols commented on src/main/java/zero/ui/Ui.java:
```diff
@@ -0,0 +1,152 @@
+     *
+     * @param task The task that was unmarked.
+     */
+    public void showTaskUnmarked(Task task) {
```
Naming of methods is nice and consistent, as well as, following the convention of being verbs! Well done!

bigismols commented on src/main/java/zero/util/Parser.java:
```diff
@@ -0,0 +1,211 @@
+     * @throws ZeroException if the command is not recognised
+     */
+    public static void parseCommand(String fullCommand, TaskList tasks, Ui ui) {
+        try {
+            if (fullCommand.equals("bye")) {
+                ui.showGoodbye();
+            } else if (fullCommand.equals("list")) {
+                ui.listTasks(tasks);
+            } else if (fullCommand.startsWith("delete")) {
+                handleDelete(tasks, fullCommand, ui);
+            } else if (fullCommand.startsWith("mark")) {
+                handleMark(tasks, fullCommand, ui);
+            } else if (fullCommand.startsWith("unmark")) {
+                handleUnmark(tasks, fullCommand, ui);
+            } else if (fullCommand.startsWith("todo")) {
+                handleTodo(tasks, fullCommand, ui);
+            } else if (fullCommand.startsWith("deadline")) {
+                handleDeadline(tasks, fullCommand, ui);
+            } else if (fullCommand.startsWith("event")) {
+                handleEvent(tasks, fullCommand, ui);
+            } else if (fullCommand.startsWith("find")) {
+                handleFind(tasks, fullCommand, ui);
+            }   else {
+                throw new ZeroException("分かりません");
+            }
+        } catch (ZeroException e) {
+            ui.showError(e.getMessage());
+        }
```
Would this be neater with switch statements?

bigismols reviewed (commented):
Generally, pretty well written code. I really like the abstraction of Zero and I like the chatbot's name too. Just be careful of naming those constants!

ywllowsensor commented on src/main/java/zero/util/Storage.java:
```diff
@@ -0,0 +1,93 @@
+        this.filePath = filePath;
+    }
+
+    /**
+     * Loads tasks from the specified file and returns them as an {@code ArrayList<Task>}.
+     * If the file does not exist, it creates the file and any necessary directories.
+     *
+     * @return An {@code ArrayList<Task>} containing the loaded tasks.
+     * @throws ZeroException If an error occurs while loading tasks.
+     */
```
I really like how descriptive and in-depth your Javadoc comments are! This makes it very simple for a fellow developer/reviewer to understand the code you are writing. Great work!

ywllowsensor reviewed (commented):
LGTM. Overall, your code is very clean and pleasant to read. Great job!

JovenSoh committed:
Add GUI to chatbot

JovenSoh committed:
Add assert to chatbot

- Ensures that the full command is not null
- Ensures that the full command is not empty
- Raises an issue if command is either of the above

JovenSoh committed:
Fix code quality

JovenSoh committed:
Fix minor bug

JovenSoh committed:
Add stream to read files

JovenSoh committed:
Set up CI

JovenSoh committed:
Edit CI configurations

JovenSoh committed:
Merge pull request #1 from JovenSoh/main

Main

JovenSoh committed:
Merge pull request #2 from JovenSoh/master

Merge 'master' into 'branch-A-CodeQuality'

JovenSoh committed:
Merge branch 'master' into branch-A-Streams

Merge 'master' into branch-A-Streams

JovenSoh committed:
Merge branch 'master' into branch-A-Streams

JovenSoh committed:
Add snooze option

This commit allows more flexibility in handling deadlines of added tasks
in a simple manner.

- Typing 'Snooze [item number]' would delay the item deadline by one day
- Handle case where item is not a deadline

JovenSoh committed:
Add snooze option

JovenSoh committed:
Add resize to input box

JovenSoh committed:
Merge branch 'master' of https://github.com/jovensoh/ip

JovenSoh committed:
Add more test cases

JovenSoh committed:
Add javadocs using AI

JovenSoh committed:
Add more error handling using AI

JovenSoh committed:
Add a User Guide

JovenSoh committed:
Update Ui.png

JovenSoh committed:
Fix documentation

JovenSoh committed:
Update README.md

JovenSoh committed:
Update README.md
